File: WorstFitMemoryAllocation.java
import java.util.ArrayList;
import java.util.Scanner;

class Block {
    int size;
    boolean isOccupied;
    int processId;

    public Block(int size) {
        this.size = size;
        this.isOccupied = false;
        this.processId = -1; // -1 indicates no process is allocated
    }
}

class WorstFitMemoryAllocation {
    ArrayList<Block> blocks;

    public WorstFitMemoryAllocation(ArrayList<Integer> blockSizes) {
        blocks = new ArrayList<>();
        for (int size : blockSizes) {
            blocks.add(new Block(size));
        }
    }

    public void allocateMemory(int processId, int processSize) {
        int worstFitIndex = -1;
        int maxBlockSize = -1;

        // Find the largest free block that can fit the process
        for (int i = 0; i < blocks.size(); i++) {
            Block block = blocks.get(i);
            if (!block.isOccupied && block.size >= processSize && block.size > maxBlockSize) {
                worstFitIndex = i;
                maxBlockSize = block.size;
            }
        }

        if (worstFitIndex != -1) {
            Block block = blocks.get(worstFitIndex);
            block.isOccupied = true;
            block.processId = processId;
            System.out.println("Allocated process " + processId + " of size " + processSize + " to block of size " + block.size);
        } else {
            System.out.println("No suitable block found for process " + processId + " of size " + processSize);
        }
    }

    public void freeMemory(int processId) {
        for (Block block : blocks) {
            if (block.isOccupied && block.processId == processId) {
                block.isOccupied = false;
                block.processId = -1;
                System.out.println("Freed memory for process " + processId);
                return;
            }
        }
        System.out.println("Process " + processId + " not found in memory.");
    }

    public void displayMemoryStatus() {
        System.out.println("\nMemory Status:");
        for (int i = 0; i < blocks.size(); i++) {
            Block block = blocks.get(i);
            String status = block.isOccupied ? "Occupied by process " + block.processId : "Free";
            System.out.println("Block " + i + " (Size: " + block.size + "): " + status);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of memory blocks: ");
        int numBlocks = scanner.nextInt();
        ArrayList<Integer> blockSizes = new ArrayList<>();

        System.out.println("Enter the sizes of each memory block:");
        for (int i = 0; i < numBlocks; i++) {
            System.out.print("Size of block " + (i + 1) + ": ");
            blockSizes.add(scanner.nextInt());
        }

        WorstFitMemoryAllocation memory = new WorstFitMemoryAllocation(blockSizes);

        while (true) {
            System.out.println("\nChoose an operation:");
            System.out.println("1. Allocate Memory");
            System.out.println("2. Free Memory");
            System.out.println("3. Display Memory Status");
            System.out.println("4. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter the process ID to allocate: ");
                    int processId = scanner.nextInt();
                    System.out.print("Enter the size of the process: ");
                    int processSize = scanner.nextInt();
                    memory.allocateMemory(processId, processSize);
                    break;
                case 2:
                    System.out.print("Enter the process ID to free: ");
                    int pid = scanner.nextInt();
                    memory.freeMemory(pid);
                    break;
                case 3:
                    memory.displayMemoryStatus();
                    break;
                case 4:
                    System.out.println("Exiting...");
                    scanner.close();
                    return;
                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }
    }
}

File: WorstFitMemoryAllocation.py
class Block:
    def __init__(self, size):
        self.size = size
        self.is_occupied = False
        self.process_id = None

class WorstFitMemoryAllocation:
    def __init__(self, block_sizes):
        self.blocks = [Block(size) for size in block_sizes]

    def allocate_memory(self, process_id, process_size):
        # Find the largest available block that can fit the process
        worst_fit_index = -1
        max_block_size = -1

        for i, block in enumerate(self.blocks):
            if not block.is_occupied and block.size >= process_size and block.size > max_block_size:
                worst_fit_index = i
                max_block_size = block.size

        if worst_fit_index != -1:
            block = self.blocks[worst_fit_index]
            block.is_occupied = True
            block.process_id = process_id
            print(f"Allocated process {process_id} of size {process_size} to block of size {block.size}")
        else:
            print(f"No suitable block found for process {process_id} of size {process_size}")

    def free_memory(self, process_id):
        # Free the block occupied by the process with the given process_id
        for block in self.blocks:
            if block.is_occupied and block.process_id == process_id:
                block.is_occupied = False
                block.process_id = None
                print(f"Freed memory for process {process_id}")
                return
        print(f"Process {process_id} not found in memory.")

    def display_memory_status(self):
        print("\nMemory Status:")
        for i, block in enumerate(self.blocks):
            status = f"Occupied by process {block.process_id}" if block.is_occupied else "Free"
            print(f"Block {i} (Size: {block.size}): {status}")

def main():
    # Get memory block sizes from the user
    block_sizes = []
    num_blocks = int(input("Enter the number of memory blocks: "))
    
    print("Enter the sizes of each memory block:")
    for i in range(num_blocks):
        block_size = int(input(f"Size of block {i + 1}: "))
        block_sizes.append(block_size)

    memory = WorstFitMemoryAllocation(block_sizes)

    # User menu for memory operations
    while True:
        print("\nChoose an operation:")
        print("1. Allocate Memory")
        print("2. Free Memory")
        print("3. Display Memory Status")
        print("4. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            process_id = int(input("Enter the process ID to allocate: "))
            process_size = int(input("Enter the size of the process: "))
            memory.allocate_memory(process_id, process_size)

        elif choice == 2:
            process_id = int(input("Enter the process ID to free: "))
            memory.free_memory(process_id)

        elif choice == 3:
            memory.display_memory_status()

        elif choice == 4:
            print("Exiting...")
            break

        else:
            print("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()
